### PR TITLE
Multi template (WIP)

### PR DESCRIPTION
**Pull request by ddooley**
Just for preliminary comments.  May hide Template menu until other templates are ready.

### CODE REVIEW

**Review by ivansg44 — CHANGES_REQUESTED** [review]
I appreciate the effort that went into this PR, and the template switching functionality seems to be working well (although there are a couple things I think you could do to improve it, mentioned at the end of this review). However, I believe more existing code is being modified in this PR than necessary. The codebase is relatively mature, so we should focus on a couple things:

1) One PR for one thing (i.e., avoid introducing sweeping changes in a single PR)
2) Extending--not modifying--the existing code

With regards to point #1, I believe we should discuss this URL parameter business, and implement it if necessary in another PR. We should just focus on UI-driven template switching in this one. I'm not convinced we need to worry about URL parameters for an application that is meant to be run locally.

With regards to point #2, I believe the point of entry in this function (you refer to it in the code as "application launch") has been modified more than necessary. Modifying existing code in a relatively mature codebase is always risky, because you can accidentally introduce bugs and break previously stable functionality. It also makes it quite difficult to review the changes, when you're also introducing a new functionality on top of this re-structuring. If there is a need to further modularize the point-of-entry, I think it should be strongly justified and in a separate PR, where we can test it thoroughly.

Here's how I would have done things:

- New JSON file
	- Keys = template names
	- Values = path to data and export
- windows.TEMPLATES = reference to data in json file, called in document.ready()
- window.TEMPLATE = (some function), called in document.ready()
	- This function takes two parameters (template name, and TEMPLATES)
	- This function returns the necessary DATA and EXPORT variables
- (some function) is also called later in document.ready(), in a codeblock similar to the ones that are already there for existing parts of the upper toolbar
	- followed by HOT.render()

Essentially, the existing code remains largely the same, with a few additions to document.ready(), and the function(s) related to switching templates kept separate from the existing code. I understand that Settings->Jump to... needs to change with the template, but I think that can be done simply by adding some sort of hook.

Couple of things I would to do to improve template switching:

1) Some sort of warning when users switch templates, if there is data in the grid
2) Move the option to switch templates under settings in the toolbar, rather than add it on the top-left

With regards to #1, see #clear-data-warning-modal and #save-as-invalid-warning-modal for precedent.

With regards to #2, my justification is based on two points. First, it keeps all the screenshots/gifs in existing documentation similar to the current UI. We can add new screenshots/gifs for this new functionality easily, but it would be a bigger pain to go change everything from previous versions. Second, it keeps the UI consistent with other spreadsheet editors, which was what I envisioned when designing this application. They usually have a tool bar on the top, with File on the left, then settings, and then help.

I do think it is important to keep the name of the current template visible to the user at all times though. Maybe the name of the current template could be on the bottom right, below the grid. That's not a big change from interfaces in our previous releases.

One last thing: I have been using the Google Style guide for HTML, CSS and JS to keep the codebase relatively uniform and consistent. I think you should follow this guide and any precedents set by the existing code, to keep the code easy-to-navigate for us and future collaboraters. Some issues I noticed were:

- Not using arrow functions, or declaring functions as constants
- Using var, instead of let and const
- Lines past the length limit
- If-statements across 2 lines without curly braces
- Uneven white-space between functions

File: main.css [hunk 1]
```diff
@@ -111,7 +111,7 @@
 #grid table.htCore > tbody > tr td:nth-child(2) {
 	border-right: medium solid gray;
 }
-
+#add-rows-input {max-width:5rem}
```
[the remark this answers is not in the document]
Comment (ddooley) [reply main.css:114]: The adjustment was not to the grid but to the element inside the grid.

File: main.html [hunk 2]
```diff
@@ -70,7 +81,7 @@
             </div>
             <input type="text" class="form-control" id="add-rows-input" value="100">
             <div class="input-group-append">
-              <span class="input-group-text">more rows at the bottom.</span>
+              <span class="input-group-text"> rows at the bottom.</span>
```
[the remark this answers is not in the document]
Comment (ddooley) [reply main.html:84]: Ok

File: script/main.js [hunk 3]
```diff
@@ -4,13 +4,24 @@
...
+};
+
 /**
  * Controls what dropdown options are visible depending on grid settings.
  */
 const toggleDropdownVisibility = () => {
   $('.hidden-dropdown-item').hide();
 
-  $('#file-dropdown-btn-group')
+  $('#file-dropdown-btn-group').off()
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:24]: Yes.  Any function call which adds event handlers to HTML elements, if the elements aren't themselves destroyed, and the function call is repeated, causes duplicate handlers, which may be bad for performance, and can lead to other bad stuff.

File: script/main.js [hunk 4]
```diff
@@ -902,12 +914,71 @@ const getComment = (field) => {
   return ret;
 };
 
+
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:917]: Ok

File: script/main.js [hunk 5]
```diff
@@ -902,12 +914,71 @@ const getComment = (field) => {
...
+/**
+ * Enable template folder's data.js vocabulary to be loaded dynamically.
+ */
+dataOnload = function () {
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:921]: good catch; I've done it on my side.

File: script/main.js [hunk 6]
```diff
@@ -902,12 +914,71 @@ const getComment = (field) => {
...
 };
 
+
+/**
+ * Enable template folder's data.js vocabulary to be loaded dynamically.
+ */
+dataOnload = function () { 
+  runBehindLoadingScreen(launch, [DATA])
+};
+
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:924]: Originally it was a larger function but now that it is winnowed down, sure.

File: script/main.js [hunk 7]
```diff
@@ -902,12 +914,71 @@ const getComment = (field) => {
...
+/**
+ * Enable template folder's export.js export options to be loaded dynamically.
+ */
+exportOnload = function () {
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:928]: done on my side.

File: script/main.js [hunk 8]
```diff
@@ -902,12 +914,71 @@ const getComment = (field) => {
...
+  const select = $("#export-to-format-select")[0];
+  while (select.options.length > 1) {
+    select.remove(1);
+  }
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:932]: getting rid of old selection options in html element before adding new ones.

File: script/main.js [hunk 9]
```diff
@@ -902,12 +914,71 @@ const getComment = (field) => {
...
+  for (const option in EXPORT_FORMATS) {
+    select.append(new Option(option,option))
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:934]: There will never be long lists of export formats beyond 5 or 10 elements say, so performance won't be a factor.

File: script/main.js [hunk 10]
```diff
@@ -902,12 +914,71 @@ const getComment = (field) => {
...
+  }
+};
+
+/************************** APPLICATION LAUNCH ********************/
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:938]: I was finding it difficult to scroll down and see where exactly the application launch section started so I emphasized it more.

File: script/main.js [hunk 11]
```diff
@@ -902,12 +914,71 @@ const getComment = (field) => {
...
+  else {//low-tech way:
+    template = location.search.split("template=")[1] || template;
+  }
+  // Try triggering template menu option so it is selected
+  if (template in TEMPLATES) {
+    $('#select-template').val(template).trigger('change');
+    return;
+  }
+  else {
+    alert (`Template "${template}" is not listed in DataHarmonizer so might not load! `);
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:960]: It was a quick message that would almost never appear for users unless we hardcoded a menu item that didn't have a template for it.  So I wasn't factoring user-facing appearance.  But I'll change it to a modal if you want.

File: script/main.js [hunk 12]
```diff
@@ -981,13 +1053,8 @@ $(document).ready(() => {
...
     }
-    if (exportFormat === 'gisaid') {
-      exportGISAID(baseName, HOT, DATA, XLSX);
-    } else if (exportFormat === 'irida') {
-      exportIRIDA(baseName, HOT, DATA, XLSX);
-    } else if (exportFormat === 'laser') {
-      exportLASER(baseName, HOT, DATA, XLSX);
-    }
+    if (exportFormat in EXPORT_FORMATS) 
+      EXPORT_FORMATS[exportFormat](baseName, HOT, DATA, XLSX);
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:1057]: done

File: script/main.js [hunk 13]
```diff
@@ -1014,6 +1108,58 @@ $(document).ready(() => {
     runBehindLoadingScreen(changeRowVisibility, args);
   });
 
+}
+
+/**
+ * Revise user interface elements to match template path, and trigger
+ * load of data.js and export.js scripts.  data_script.onload goes on
+ * to trigger launch(DATA).
+ * @param {String} template: path of template starting from app's template folder.
```
[the remark this answers is not in the document]
Comment (ddooley) [reply script/main.js:1117]: done

**Comment by ddooley** [comment]
This pr I should have marked WIP. Its meant to be a design iteration to be discussed. I wrote it in a day.
- I tried to follow your / Google code pattern but neglected a few things here and there, so thanks for noting those spots.
- Re. modal - the alert isn't something a user would see normally; it was more a programmer level message. but I've done the change.
- Re. template via URI. Its a feature I popped in there to show how we can handle templates that don't necessarily appear in menu system.
- Re. display/selection of templates in menu vs in setup vs. title somewhere on screen.  Totally a point of discussion as I mentioned in the pr. A decision we can all dwell over.
- Re. mature code base. The main problem with the current code base was that the startup part had a variety of jquery html handler connections that are only run once, mixed in with html setup stuff that needed to be rerun on each template load. So I separated those out.  Secondly I had to split into a few stages the code that loads the template json/js and then do stuff after the load.  I think you would discover that you would have to do the very same things, so it wasn't a capricious reorganization.
- As to the suggestion of a TEMPLATE data structure, with functions inside it and data.js too, I agree, we can head in that kind of direction.  That exposes that the code could be refactored in a more modular way as a class with data components, initializers etc. I didn't have time to go that far.

**Comment by ddooley** [comment]
Note that some data structures and functions require "var" definitions in certain cases for objects and functions where js scripts are reloaded, otherwise javascript throws an error

**Comment by ddooley** [comment]
I've changed the template selection menu to be in a modal, which also enables another checkbox option to view under-development templates as well.

**Pull request merged**